segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/SyncStorage.java

```diff
@@ -216,6 +216,40 @@ default SegmentHandle create(String streamSegmentName, SegmentRollingPolicy roll
      */
     boolean supportsTruncation();
 
+    /**
+     * Gets a value indicating whether this {@link SyncStorage} implementation can replace whole Segments with new contents.
+     *
+     * @return True or false.
+     */
+    default boolean supportsReplace() {
+        return false;
+    }
+
+    /**
+     * Replaces a Segment with the given contents.  Please refer to the actual implementing class for more details with
+     * respect to behavior, atomicity and recovery mechanisms.
+     *
+     * @param segment A {@link SegmentHandle} representing the Segment to replace.
+     * @param contents A {@link BufferView} representing the new contents of the Segment.
+     * @throws StreamSegmentNotExistsException When the given Segment does not exist in Storage.
+     * @throws StorageNotPrimaryException      When this Storage instance is no longer primary for this Segment (it was
+     *                                         fenced out).
+     * @throws UnsupportedOperationException   If {@link #supportsReplace()} returns false.
+     */
+    default void replace(SegmentHandle segment, BufferView contents) throws StreamSegmentException {
```

Comment:
But I am not renaming a file here. What I need is to set the contents of a Segment to a given buffer. I have that buffer in `contents` and this method is expected to set that as the actual contents of `segment`. Whether or not a `rename` operation happens behind the scenes to implement this is a different story.

For example, in `ExtendedS3Storage` this may be accomplished by a `putObjectRequest` that overwrites a target object - no need for rename.

For this fix, we do not need or care about renaming segments. We care about "atomically" replacing the contents of a segment with some other contents that we have already generated (in memory).